execution_engine/src/core/runtime/host_function_flag.rs

```diff
@@ -0,0 +1,95 @@
+use std::{cell::Cell, rc::Rc};
+
+use tracing::error;
+
+/// A flag to indicate whether the current runtime call is made within the scope of a host function.
+///
+/// The flag is backed by an `Rc<Cell<u64>>`, meaning that clones will all share state.
+#[derive(Default, Clone)]
+pub(super) struct HostFunctionFlag {
+    /// A counter which, if non-zero, indicates that the `HostFunctionFlag` is `true`.
+    counter: Rc<Cell<u64>>,
+}
+
+impl HostFunctionFlag {
+    /// Returns `true` if this `HostFunctionFlag` has entered any number of host function scopes
+    /// without having exited them all.
+    pub(super) fn is_in_host_function_scope(&self) -> bool {
+        self.counter.get() != 0
+    }
+
+    /// Must be called when entering a host function scope.
+    ///
+    /// The returned `ScopedHostFunctionFlag` must be kept alive for the duration of the host
+    /// function call.  While at least one such `ScopedHostFunctionFlag` exists,
+    /// `is_in_host_function_scope()` returns `true`.
+    #[must_use]
+    pub(super) fn enter_host_function_scope(&self) -> ScopedHostFunctionFlag {
+        let new_count = self.counter.get().checked_add(1).unwrap_or_else(|| {
+            error!("checked_add failure in host function flag counter");
+            u64::MAX
+        });
+        self.counter.set(new_count);
+        ScopedHostFunctionFlag {
+            counter: self.counter.clone(),
+        }
+    }
+}
+
+pub(super) struct ScopedHostFunctionFlag {
+    counter: Rc<Cell<u64>>,
+}
+
+impl Drop for ScopedHostFunctionFlag {
+    fn drop(&mut self) {
+        let new_count = self.counter.get().checked_sub(1).unwrap_or_else(|| {
+            error!("checked_sub failure in host function flag counter");
```

Comment:
Same, panicking is better because the counter being zero here means we dropped more flags than we created.

Reply:
Again, I don't believe it's possible to hit this path.  If we do, the repercussions are less severe, in that a call to a host function might be overcharged.  But the node and network will continue to run.

We do treat error level log lines as critical issues, so if this were to appear in a logfile from testing or elsewhere, it would be investigated.

Reply:
I would have to disagree here. The value being zero before a `drop` means we have violated Rust's object model, which would be a high severity issue. It would have to be caused by, for instance, funny pointer business in unsafe code or memory bit flips. In such a situation, we wouldn't want to continue execution with such an unreliable node: it might lead to forks.